### Проект по SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. 
В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. 
Зато стало больше времени для книг. 

Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 
Ваша первая задача как аналитика — проанализировать базу данных. 
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Описание данных

Таблица books
Содержит данные о книгах:
* book_id — идентификатор книги;
* author_id — идентификатор автора;
* title — название книги;
* num_pages — количество страниц;
* publication_date — дата публикации книги;
* publisher_id — идентификатор издателя.

Таблица authors
* Содержит данные об авторах:
* author_id — идентификатор автора;
* author — имя автора.

Таблица publishers
* Содержит данные об издательствах:
* publisher_id — идентификатор издательства;
* publisher — название издательства;

Таблица ratings
* Содержит данные о пользовательских оценках книг:
* rating_id — идентификатор оценки;
* book_id — идентификатор книги;
* username — имя пользователя, оставившего оценку;
* rating — оценка книги.

Таблица reviews
* Содержит данные о пользовательских обзорах:
* review_id — идентификатор обзора;
* book_id — идентификатор книги;
* username — имя автора обзора;
* text — текст обзора.

### Задания

* Посчитайте, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитайте количество обзоров и среднюю оценку;
* Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
* Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
* Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Цели исследования: 

Первичный анализ базы данных сервиса для чтения книг по подписке. Анализ информации о книгах, издательствах, авторах, а также пользовательских обзорах книг для формулирования в дальнейшем ценностного предложения для нового продукта.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
# импортируем каждую таблицу в отдельный датафрейм
query_books = ''' SELECT * FROM books; ''' 
books = pd.io.sql.read_sql(query_books, con = engine)

query_authors = ''' SELECT * FROM authors; ''' 
authors = pd.io.sql.read_sql(query_authors, con = engine)

query_publishers = ''' SELECT * FROM publishers; ''' 
publishers = pd.io.sql.read_sql(query_publishers, con = engine)

query_ratings = ''' SELECT * FROM ratings; ''' 
ratings = pd.io.sql.read_sql(query_ratings, con = engine)

query_reviews = ''' SELECT * FROM reviews; ''' 
reviews = pd.io.sql.read_sql(query_reviews, con = engine)

In [3]:
# функция для исследовательского анализа данных 
def info_data(a):
    display(a.head()),
    a.info(),
    print('Дубликатов',a.duplicated().sum()),
    display(a.describe())

### Проанализируем полученные датафреймы

In [4]:
info_data (books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
Дубликатов 0


book_id    author_id   num_pages  publisher_id
count  1000.000000  1000.000000  1000.00000   1000.000000
mean    500.500000   320.417000   389.11100    171.270000
std     288.819436   181.620172   229.39014     99.082685
min       1.000000     1.000000    14.00000      1.000000
25%     250.750000   162.750000   249.00000     83.000000
50%     500.500000   316.500000   352.00000    177.500000
75%     750.250000   481.000000   453.00000    258.000000
max    1000.000000   636.000000  2690.00000    340.000000

In [5]:
info_data (authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
Дубликатов 0


author_id
count  636.000000
mean   318.500000
std    183.741666
min      1.000000
25%    159.750000
50%    318.500000
75%    477.250000
max    636.000000

In [6]:
info_data (publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
Дубликатов 0


publisher_id
count    340.000000
mean     170.500000
std       98.293777
min        1.000000
25%       85.750000
50%      170.500000
75%      255.250000
max      340.000000

In [7]:
info_data (ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
Дубликатов 0


rating_id      book_id       rating
count  6456.000000  6456.000000  6456.000000
mean   3228.500000   510.574195     3.928284
std    1863.831001   284.141636     0.943303
min       1.000000     1.000000     1.000000
25%    1614.750000   291.000000     3.000000
50%    3228.500000   506.000000     4.000000
75%    4842.250000   750.000000     5.000000
max    6456.000000  1000.000000     5.000000

In [8]:
info_data (reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
Дубликатов 0


review_id      book_id
count  2793.000000  2793.000000
mean   1397.000000   504.693161
std     806.413976   288.472931
min       1.000000     1.000000
25%     699.000000   259.000000
50%    1397.000000   505.000000
75%    2095.000000   753.000000
max    2793.000000  1000.000000

* Дублкатов в данных не обнаружено
* Пропуски данных не обнаружены
* С типами данных все в порядке, за исключением столбца publication_date таблицы books - для него требуется замена формата времени

### Посчитаем, сколько книг вышло после 1 января 2000 года

In [9]:
query_date = ''' 
SELECT COUNT(*) 
FROM books
WHERE publication_date > '2000-1-1';; 
''' 
count_date = pd.io.sql.read_sql(query_date, con = engine)

print(f'После 1 января 2000 года вышло {count_date.iloc[0,0]} книг')

После 1 января 2000 года вышло 819 книг


* После 1 января 2000 года вышло 819 книг

### Посчитаем для каждой книги количество обзоров и среднюю оценку

In [10]:
query_rating = ''' 

SELECT 
    b.title AS book_name,
    b.book_id AS id,
    rew_new.new_reviews AS reviews,
    r_new.avg_rating AS avg_rating
FROM books AS b LEFT JOIN 
                (SELECT
                    book_id,
                    AVG(rating) AS avg_rating
                FROM
                    ratings 
                GROUP BY 
                    book_id
                ) AS r_new ON r_new.book_id = b.book_id
        LEFT JOIN
                (SELECT 
                    book_id,
                    COUNT(review_id) AS new_reviews
                FROM
                    reviews
                GROUP BY 
                    book_id
                ) AS rew_new ON rew_new.book_id = b.book_id
                
; ''' 

request_rating = pd.io.sql.read_sql(query_rating, con = engine)
request_rating.head()

book_name   id  reviews  avg_rating
0          The Body in the Library (Miss Marple  #3)  652      2.0    4.500000
1                                          Galápagos  273      2.0    4.500000
2                           A Tree Grows in Brooklyn   51      5.0    4.250000
3  Undaunted Courage: The Pioneering First Missio...  951      2.0    4.000000
4                                        The Prophet  839      4.0    4.285714

In [11]:
request_rating['id'].count()

1000

* Составили таблицу с указанием количества обзоров и средней оценкой для каждой книги

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц

In [12]:
query_publisher = ''' 
SELECT
    S.publisher_id,
    S.publisher,
    S.books_count
FROM
    (SELECT 
        b.publisher_id AS publisher_id,
        COUNT(b.book_id) AS books_count,
        p.publisher AS publisher 
    FROM 
        books AS b
        LEFT JOIN publishers AS p ON b.publisher_id=p.publisher_id

    WHERE 
        b.num_pages > 50
    GROUP BY
        b.publisher_id,
        p.publisher
    ORDER BY 
        books_count DESC) AS S
LIMIT 1
; ''' 

request_publisher = pd.io.sql.read_sql(query_publisher, con = engine)
request_publisher

publisher_id      publisher  books_count
0           212  Penguin Books           42

* Наибольшее чисо книг (42 шт.) толще 50 страниц выпустило издательство Penguin Books

### Определим автора с самой высокой средней оценкой книг 

Учитываются только книги с 50 и более оценками

In [13]:
query_author = ''' 
SELECT 
    author_rating.author,
    AVG(author_rating.avg_rating) AS avg_rating
FROM
        (SELECT 
            a.author AS author, 
            AVG(r.rating) AS avg_rating
        FROM books AS b
            LEFT JOIN authors AS a ON b.author_id = a.author_id
            LEFT JOIN ratings AS r ON b.book_id = r.book_id
        GROUP BY 
             b.book_id,a.author_id
        HAVING
            COUNT(r.rating_id) >=50 
        ) AS author_rating
GROUP BY    
    author_rating.author
ORDER BY
    avg_rating DESC
LIMIT 1
    
; ''' 

author = pd.io.sql.read_sql(query_author, con = engine)
author

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

* Автор с самой высокой средней оценкой книг - J.K. Rowling с иллюстрациями Mary GrandPré
* Средний рейтинг 4,28

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [14]:
query_reviews = ''' 

SELECT
    AVG(r.count_reviews)

FROM
    (SELECT username,
            COUNT(review_id) AS count_reviews
    FROM reviews
    WHERE username IN 
        
        (SELECT username
         FROM ratings
         GROUP BY username
            HAVING COUNT(rating_id) > 50)
    GROUP BY username
    ) AS r
; '''

reviews = pd.io.sql.read_sql(query_reviews, con = engine)
reviews

avg
0  24.333333

* Среднее количество обзоров от пользователей, которые поставили более 50 оценок: 24,3